test/Common.FunctionalTests/AppOfflineTests.cs

```diff
@@ -29,32 +30,114 @@ public class AppOfflineTests : IISFunctionalTestBase
         }
 
-        [ConditionalTheory]
-        [InlineData(HostingModel.InProcess)]
-        [InlineData(HostingModel.OutOfProcess)]
-        public async Task AppOfflineDroppedWhileSiteIsDown_CustomResponse(HostingModel hostingModel)
+        [ConditionalFact]
+        public async Task AppOfflineDroppedWhileSiteFailedToStartInShim_AppOfflineServed_InProcess()
         {
-            var expectedResponse = "The app is offline.";
-            var deploymentResult = await DeployApp(hostingModel);
+            var deploymentResult = await DeployApp(HostingModel.InProcess);
+
+            Helpers.ModifyAspNetCoreSectionInWebConfig(deploymentResult, "processPath", "nonexistent");
+
+            var result = await deploymentResult.HttpClient.GetAsync("/");
+            Assert.Equal(500, (int)result.StatusCode);
+            Assert.Contains("500.0", await result.Content.ReadAsStringAsync());
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+
+            await AssertAppOffline(deploymentResult);
+            AssertFilesNotLocked(deploymentResult);
+        }
+
+        [ConditionalFact]
+        [RequiresIIS(IISCapability.ShutdownToken)]
+        public async Task AppOfflineDroppedWhileSiteFailedToStartInRequestHandler_SiteStops_InProcess()
+        {
+            var deploymentResult = await DeployApp(HostingModel.InProcess);
 
-            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot, expectedResponse);
+            // Set file content to empty so it fails at runtime
+            File.WriteAllText(Path.Combine(deploymentResult.DeploymentResult.ContentRoot, "Microsoft.AspNetCore.Server.IIS.dll"), "");
 
-            await AssertAppOffline(deploymentResult, expectedResponse);
+            var result = await deploymentResult.HttpClient.GetAsync("/");
+            Assert.Equal(500, (int)result.StatusCode);
+            Assert.Contains("500.30", await result.Content.ReadAsStringAsync());
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+            AssertStopsProcess(deploymentResult);
+        }
+
+        
+        [ConditionalFact]
+        [RequiresIIS(IISCapability.ShutdownToken)]
+        public async Task AppOfflineDroppedWhileSiteFailedToStart_SiteStops_OutOfProcess()
+        {
+            var deploymentResult = await DeployApp(HostingModel.OutOfProcess);
+
+            Helpers.ModifyAspNetCoreSectionInWebConfig(deploymentResult, "processPath", "nonexistent");
+
+            var result = await deploymentResult.HttpClient.GetAsync("/");
+            Assert.Equal(502, (int)result.StatusCode);
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+            await AssertAppOffline(deploymentResult);
+            AssertFilesNotLocked(deploymentResult);
         }
 
         [ConditionalFact]
+        [RequiresIIS(IISCapability.ShutdownToken)]
+        public async Task AppOfflineDroppedWhileSiteStarting_SiteShutsDown_InProcess()
```

Comment:
I like the idea of this test, but from reading the implementation I'm afraid it might be too unreliable.  Does the test depend on hitting a certain race condition within 10 tries (the `for` loop), and if it doesn't then the test fails?  Are we confident this won't become a flaky test?

Reply:
> Does the test depend on hitting a certain race condition within 10 tries 

Yes. Unfortunately, this is as close as I can get to reproing the issue. Additionally considering how long it takes the site to start it should be hitting the race pretty frequently.

>  Are we confident this won't become a flaky test

No, but I'd like to test this scenario. If test becomes flaky I'll try to harden it even more.